### PR TITLE
Add keyboard shorcuts to focus and unfocus chat

### PR DESCRIPTION
**Pull request by Xenos6666**
## Short roundup of the initial problem
It isn't possible to focus and unfocus the in-game chat with keyboard shortcuts only

## What will change with this Pull Request?
- A new shortcut has been added to focus the chat box in the game tab
- A new shortcut has been added that unfocus any line input in the main window

_(Unrelated: added include sorting to clang-format configuration)_

### CODE REVIEW

**Comment by Xenos6666** [comment]
I'm not very familiar with the Qt library, and for some reason I needed to add the shortcut action to a menu for it to work. Is there a way to make it work that doesn't require to add it to a menu? This could also be useful for future keyboard control support I plan to work on

**Comment by ebbit1q** [comment]
It should be, I prefer to just add it to a menu as well though.

**Comment by ebbit1q** [comment]
Just checked, ctrl e after clicking chat doesn't work on master either. This is in a local game.

**Comment by Xenos6666** [comment]
Unfocusing the chat worked for me when I tested it in a local game, I'll be testing it further later today when I have the time, I might also take a look at the Ctrl+e bug while I'm at it

**Comment by Xenos6666** [comment]
I did some tests on the Ctrl+e error, it seem like keys that already have a function in chat don't trigger the associated shortcuts, which is the expected behavior (Ctrl+e goes to the end of line), but it's not really a problem, since the main aim of adding the escape behavior was to prevent these conflicts (it also allows people to have key bindings without Ctrl). However, by, fooling around, I found something else, it seems like shortcuts set by the player.cpp file don't work with multiple bindings on the same shortcut, the problem seems to come from the `setShortcutsActive` function, I'm not sure if this is intended behavior.

**Comment by Xenos6666** [comment]
As for the Esc bug, it seems to work as long as the game hasn't started, but stops working after that. I think I'll try implementing it on each instances of LineEditCompleter to make it more stable and prevent problems with the focus being stuck in other parts of the application (for example in the deck editor tab's search bar)

**Comment by Xenos6666** [comment]
Actually, I just noticed that Search bar isn't an instance of LineEditCompleter

**Comment by Xenos6666** [comment]
I had 2 alternatives to have get the behavior I wanted : 
- Catching the Esc key press on the MainWindow, and checking the currently focused Widget
- Catching the Esc key press on the QLineEdit widget itself

The first one had the potential to create bugs and conflicts with other people's codes, while the second one is probably less error prone but required a minor restructuring of many disjointed places of the code, and could cause inconsistencies in future codes. I chose to go with the second one, but if you prefer the first option it wouldn't take me long to re-code it.

**Comment by Xenos6666** [comment]
Wait, the fix didn't actually work, but now I know what causes it, the unfocusing doesn't work if a shortcut is set using the esc key

**Comment by Xenos6666** [comment]
All issues should be fixed

File: cockatrice/src/customlineedit.h [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+#include <QLineEdit>
+
+class QKeyEvent;
+class QWidget;
+class QString;
+
+// Should be used when the there is a risk of conflict between line editor
+// shortcuts and other shortcuts
+class CustomLineEdit : public QLineEdit
```
Comment (ebbit1q) [inline cockatrice/src/customlineedit.h]: CustomLineEdit is to vague in my opinion, this is specifically a lineedit that you can unfocus with a key.

File: cockatrice/src/lineeditcompleter.h [hunk 2]
```diff
@@ -3,10 +3,11 @@
 
 #include <QFocusEvent>
 #include <QKeyEvent>
-#include <QLineEdit>
 #include <QStringList>
 
-class LineEditCompleter : public QLineEdit
+#include "customlineedit.h"
```
Comment (ebbit1q) [inline cockatrice/src/lineeditcompleter.h]: project include above others, see https://github.com/Cockatrice/Cockatrice/blob/master/.github/CONTRIBUTING.md#header-files

File: cockatrice/src/tab_deck_editor.h [hunk 3]
```diff
@@ -5,9 +5,9 @@
 #include "tab.h"
 #include <QAbstractItemModel>
 #include <QDir>
-#include <QLineEdit>
 
 #include "carddatabase.h"
```
Comment (ebbit1q) [inline cockatrice/src/tab_deck_editor.h]: this one already was at the wrong spot

File: cockatrice/src/tab_logs.cpp [hunk 4]
```diff
@@ -19,6 +14,13 @@
 #include <QtGui>
 #include <QtWidgets>
 
+#include "abstractclient.h"
```
Comment (ebbit1q) [inline cockatrice/src/tab_logs.cpp]: same

File: cockatrice/src/tab_message.cpp [hunk 5]
```diff
@@ -6,11 +6,11 @@
 #include "soundengine.h"
 #include <QApplication>
 #include <QDebug>
-#include <QLineEdit>
 #include <QMenu>
 #include <QSystemTrayIcon>
 #include <QVBoxLayout>
 
+#include "customlineedit.h"
```
Comment (ebbit1q) [inline cockatrice/src/tab_message.cpp]: oh dear, I guess we're not very consistent

File: cockatrice/src/window_sets.cpp [hunk 6]
```diff
@@ -19,6 +16,11 @@
 
 #include <algorithm>
 
+#include "customlineedit.h"
```
Comment (ebbit1q) [inline cockatrice/src/window_sets.cpp]: same

File: cockatrice/src/customlineedit.cpp [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    }
+
+    QLineEdit::keyPressEvent(event);
+}
+
+bool CustomLineEdit::eventFilter(QObject *watched, QEvent *event)
+{
+    if (event->type() == QEvent::ShortcutOverride) {
+        QKeyEvent *keyEvent = static_cast<QKeyEvent *>(event);
+        if (keyEvent->key() == Qt::Key_Escape) {
```
Comment (ebbit1q) [inline cockatrice/src/customlineedit.cpp]: Would it be possible to make this shortcut configurable?

**Comment by ZeldaZach** [comment]
I'm pretty sure the review by @ebbit1q Can be solved by running the clang-format file on the codebase

**Comment by Xenos6666** [comment]
Clang format wasn't configured to sort headers, I added a few lines to fix that. I also renamed CustomLineEdit to LineEditUnfocusable, and added a configurable shortcut for unfocusing, as requested. However, I couldn't manage to make a `refreshShortcuts()` function work on the LineEditUnfocusable for some reason, so I decided to check the shortcuts at each key presses done when the widget is focused, this seems very inefficient to me, but it's the only solution I found.

**Comment by tooomm** [comment]
>Clang format wasn't configured to sort headers, I added a few lines to fix that

Very appreciated addition, really nice!
But please do that in a separate PR next time as it is not related to this at all. 👍

**Comment by Xenos6666** [comment]
I'll keep that in mind, thanks! (It's the first time I've made a pull request, I'm learning how things work)

**Pull request merged**